### PR TITLE
Update TaskRole to access only ECS CLI resources

### PR DESCRIPTION
**Pull request by efekarakus**
Currently the TaskRole has only `PowerUserAccess`, this change restricts the permissions to:
1. Blocks IAM calls.
2. Only storage and messaging services that are prefixed with the `{project}-{env}`
3. Tagged resources

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: templates/lb-fargate-service/cf.yml [hunk 1]
```diff
@@ -118,8 +124,52 @@ Resources:
...
+              - Effect: 'Deny'
+                Action: 'iam:*'
+                Resource: '*'
+        - PolicyName: 'AllowPrefixedResources'
+          PolicyDocument:
+            Version: '2012-10-17'
+            Statement:
+              - Effect: 'Allow'
+                Action: '*'
+                Resource:
```
Comment (efekarakus) [inline templates/lb-fargate-service/cf.yml]: Let me know if you can think of other services that are likely to be accessed by application code!

File: templates/lb-fargate-service/cf.yml [hunk 2]
```diff
@@ -118,8 +124,52 @@ Resources:
...
+                  - !Sub 'arn:aws:sns:${AWS::Region}:${AWS::AccountId}:${ProjectName}-${EnvName}-*'
+                  - !Sub 'arn:aws:sqs:${AWS::Region}:${AWS::AccountId}:${ProjectName}-${EnvName}-*'
+                  - !Sub 'arn:aws:kinesis:${AWS::Region}:${AWS::AccountId}:*/${ProjectName}-${EnvName}-*'
+                  - !Sub 'arn:aws:firehose:${AWS::Region}:${AWS::AccountId}:*/${ProjectName}-${EnvName}-*'
+                  - !Sub 'arn:aws:kinesisanalytics:${AWS::Region}:${AWS::AccountId}:*/${ProjectName}-${EnvName}-*'
+        - PolicyName: 'AllowTaggedResources' # See https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_actions-resources-contextkeys.html
+          PolicyDocument:
+            Version: '2012-10-17'
+            Statement:
+              - Effect: 'Allow'
```
[the remark this answers is not in the document]
Comment (efekarakus) [reply templates/lb-fargate-service/cf.yml]: The `ResourceTag` is for accessing tagged resources, on the other hand `RequestTag` is for tagging/untagging resources

File: templates/lb-fargate-service/cf.yml [hunk 3]
```diff
@@ -118,8 +124,52 @@ Resources:
...
-        - 'arn:aws:iam::aws:policy/PowerUserAccess'
+      Policies:
+        - PolicyName: 'DenyIAMActions'
+          PolicyDocument:
+            Version: '2012-10-17'
+            Statement:
+              - Effect: 'Deny'
+                Action: 'iam:*'
+                Resource: '*'
+        - PolicyName: 'AllowPrefixedResources'
```
[the remark this answers is not in the document]
Comment (efekarakus) [reply templates/lb-fargate-service/cf.yml]: Done!

[the remark this answers is not in the document]
Comment (efekarakus) [reply templates/lb-fargate-service/cf.yml]: Added (taken from SAM), hopefully that's considered okay:
```
 Statement:
              - Effect: 'Allow'
                Action:
                  - 'cloudwatch:PutMetricData'
                Resource: '*'
              - Effect: 'Allow'
                Action:
                  - 'cloudwatch:GetDashboard'
                  - 'cloudwatch:ListDashboards'
                  - 'cloudwatch:PutDashboard'
                  - 'cloudwatch:ListMetrics'
                Resource: '*'
```

File: templates/lb-fargate-service/cf.yml (hunk 1 above)
Comment (efekarakus) [reply templates/lb-fargate-service/cf.yml]: Added (taken from SAM), hopefully that's considered okay:
```
              Statement:
              - Effect: 'Allow'
                Action:
                  - 'cloudwatch:PutMetricData'
                Resource: '*'
              - Effect: 'Allow'
                Action:
                  - 'cloudwatch:GetDashboard'
                  - 'cloudwatch:ListDashboards'
                  - 'cloudwatch:PutDashboard'
                  - 'cloudwatch:ListMetrics'
                Resource: '*'
```

File: templates/lb-fargate-service/cf.yml [hunk 4]
```diff
@@ -118,8 +124,74 @@ Resources:
...
+                  StringEquals:
+                    'aws:ResourceTag/ecs-project': !Sub '${ProjectName}'
+                    'aws:ResourceTag/ecs-environment': !Sub '${EnvName}'
+              - Effect: 'Allow'
+                Action: '*'
+                Resource: '*'
+                Condition:
+                  StringEquals:
+                    'secretsmanager:ResourceTag/ecs-project': !Sub '${ProjectName}'
+                    'secretsmanager:ResourceTag/ecs-environment': !Sub '${EnvName}'
```
Comment (sonofachamp) [inline templates/lb-fargate-service/cf.yml]: I'm assuming the Statement above this one is not covering `secretsmanager`? The only difference seems to be `aws` versus `secretsmanager`?

Comment (efekarakus) [reply templates/lb-fargate-service/cf.yml]: yeah correct, some services don't have the "aws" namespace. I'm not sure why :P I guess it's some legacy reason.

For example, [ecs](https://docs.aws.amazon.com/IAM/latest/UserGuide/list_amazonelasticcontainerservice.html#amazonelasticcontainerservice-resources-for-iam-policies) has both the aws and service namespace.  
On the other hand, [other services](https://docs.aws.amazon.com/IAM/latest/UserGuide/list_awssecretsmanager.html#awssecretsmanager-resources-for-iam-policies) don't have the "aws" namespace.

Comment (hencrice) [reply templates/lb-fargate-service/cf.yml]: Is it possible to combine the 2?

Comment (efekarakus) [reply templates/lb-fargate-service/cf.yml]: I thought that would create an AND relation and result in matching nothing.

I'll test it out and send a follow-up if it's combinable

**Pull request merged**